pkg/state/podresources/podresources.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podresources
+
+import (
+	"sync"
+
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/karpenter/pkg/utils/resources"
+)
+
+// This class is primarily used to track the overall resource requests for pods in the cluster
+// it is a proxy for the overall desired size of the cluster. It is flawed, and should be replaced
+// with a superior desired resource count mechanism that handles the following cases:
+//  1. Pod manager objects (replicasets, jobs) and their replacement workflows.
+//  2. Standalone pods
+//  3. IPPA
+
+type PodResources struct {
+	sync.RWMutex
+	podMap map[string]corev1.ResourceList
+	total  corev1.ResourceList
+	count  int
+}
+
+func NewPodResources() *PodResources {
+	return &PodResources{
+		podMap: make(map[string]corev1.ResourceList),
+		total:  nil,
```

Comment:
It's probably a good idea to initialize this to an empty resource list rather than `nil`. Right now if we call `GetTotalPodResourceRequests` before `UpdatePod` it will panic.